File: scripts/v3/modules/price/ref/CalcVolToTarget.ts
import { IRL, InRangeLiquidity, V3Reserves } from '../inRangeLiquidity';
import { LiquidityMath, Pool, SwapMath, TickMath } from "@uniswap/v3-sdk";
import { ERC20token, PoolInfo, PoolStateV3, Slot0 } from '../../../../../constants/interfaces';
import { abi as IUniswapV3PoolState } from '@uniswap/v3-core/artifacts/contracts/interfaces/pool/IUniswapV3PoolState.sol/IUniswapV3PoolState.json';
import { abi as IUniswapV3Pool } from '@uniswap/v3-core/artifacts/contracts/interfaces/IUniswapV3Pool.sol/IUniswapV3Pool.json';
import { abi as IAlgebraPool } from '@cryptoalgebra/core/artifacts/contracts/AlgebraPool.sol/AlgebraPool.json';
import { ethers, Contract } from 'ethers';
import { ExchangeMapV3, uniswapV3Exchange } from '../../../../../constants/addresses';
import { provider } from '../../../../../constants/provider';
import { pu } from '../../../../modules/convertBN';
import { error } from 'console';

export interface ITick {
	liquidityGross: bigint
	liquidityNet: bigint
	feeGrowthOutside0X128: bigint
	feeGrowthOutside1X128: bigint
	tickCumulativeOutside: bigint
	secondsPerLiquidityOutsideX128: bigint
	secondsOutside: bigint
	initialized: boolean
}

export interface ATick {
	liquidityTotal: bigint
	liquidityDelta: bigint
	outerFeeGrowth0Token: bigint
	outerFeeGrowth1Token: bigint
	prevTick: bigint
	nextTick: bigint
	outerSecondsPerLiquidity: bigint
	outerSecondsSpent: bigint
	hasLimitOrders: boolean
}
// contract - the this.pool's contract
// sCurrentPrice - sqrt of the current price
// sPriceTarget - sqrt of the target price
// liquidity - the liquidity in the current tick range of the this.pool
// tickLower, tickUpper - the min and max ticks of the current tick range
// sPriceUpper, sPriceUpper - square roots of prices corresponding to the min and max ticks of the current range
// tickSpacing - the tick spacing in the this.pool.
// 	decimalsX, decimalsY - the number of decimals of the X and Y tokens, for printing the result
export class VolToTarget {
	exchange: string
	token0: ERC20token
	token1: ERC20token
	pool: Contract
	data: InRangeLiquidity
	sPriceTarget: number

	constructor(exchange: string, token0: ERC20token, token1: ERC20token, pool: Contract, data: InRangeLiquidity, sPriceTarget: number) {
		this.exchange = exchange
		this.token0 = token0
		this.token1 = token1
		this.pool = pool
		this.data = data
		this.sPriceTarget = sPriceTarget
	}


	//  amount of x in range; 
	// sp = sPriceCurrent, sb = sPriceUpper
	x_in_range(L: number, sp: number, sb: number) {
		return L * (sb - sp) / (sp * sb)
	}

	//  amount of y in range; 
	// sp = sPriceCurrent, sa = sPriceLower
	y_in_range(L: number, sp: number, sa: number) {
		return L * (sp - sa)
	}

	tick_to_price(tick: number) {
		return 1.0001 ** tick
	}

	async calcVolToTarget(): Promise<bigint> {
		const s = await this.data.getReserves()
		const s0 = await this.data.getSlot0()
		let sPriceCurrent = Number(s.sqrtPrice)
		let sPriceLower = Number(s.sqrtRatioLow)
		let sPriceUpper = Number(s.sqrtRatioHigh)
		let liquidity = Number(s.liquidity)
		let tickLower = Number(s.tickLow)
		let tickUpper = Number(s.tickHigh)


		// //  how much of X or Y tokens we need to * buy * to get to the target price ?
		let deltaTokens = 0
		let x = 0
		let delta: bigint = 0n
		let nextTickRange: any
		let currentTickRange: any

		// The Trade class logic makes it always so that the target price is higher than the current price
		// The logic below is for the opposite case, when the target price is lower than the current price
		// if (sPriceCurrent < this.sPriceTarget) {
		// 	console.log("targetPool priceOut is already lower than the target price")
		// 	return 0n
		// }
		// IF WE WANT TO USE LIQUIDITY INSTEAD OF PRICE TO DETERMINE DIRECTION, WE CAN WRAP OUR HEADS AROUND THIS:

		//  too few Y in the this.pool; we need to buy some X to increase amount of Y in this.pool
		try {
			if (this.sPriceTarget > sPriceCurrent) {
				if (this.sPriceTarget > sPriceUpper) {
					//  not in the current price range; use all X in the range
					x = this.x_in_range(liquidity, sPriceCurrent, sPriceUpper)

					deltaTokens += x
					nextTickRange = await this.pool.ticks(BigInt(tickUpper))
					liquidity += Number(nextTickRange[1])
					// adjust the price and the range limits
					sPriceCurrent = sPriceUpper
					tickLower = tickUpper
					tickUpper += s.tickSpacing
					sPriceLower = sPriceUpper
					sPriceUpper = this.tick_to_price(tickUpper / 2)
				} else {
					// in the current price range
					x = this.x_in_range(liquidity, sPriceCurrent, this.sPriceTarget)
					deltaTokens += x
					sPriceCurrent = this.sPriceTarget

				}
				console.log("token0 tradeSize: ", (deltaTokens / 10 ** this.token0.decimals), "from ", uniswapV3Exchange[this.exchange].protocol)
				// console.log("fixedNumber: ", deltaTokens.toFixed(this.token0.decimals))
				delta = BigInt(deltaTokens)
				return delta
			}
		} catch (e: any) {
			if (e.code === 'BAD_DATA') {
				console.log("Protocol: ", uniswapV3Exchange[this.exchange].protocol)
				// console.log("Error: ", e)
				// return 0n
			}
			console.log('tickUpper: ', tickUpper)
			console.log(e.code)
		}
		try {
			if (sPriceCurrent < this.sPriceTarget) {
				// 	return 0n
				// }
				while (this.sPriceTarget < sPriceCurrent) {
					if (this.sPriceTarget < sPriceLower) {
						//  not in the current price range; use all Y in the range
						deltaTokens += this.y_in_range(liquidity, sPriceCurrent, sPriceLower)
						//  query the blockchain for liquidity in the previous tick range
						currentTickRange = await this.pool.ticks(tickLower)
						liquidity -= Number(currentTickRange[1])
						// adjust the price and the range limits
						sPriceCurrent = sPriceLower
						tickUpper = tickLower
						tickLower -= Number(s.tickSpacing)
						sPriceUpper = sPriceLower
						sPriceLower = this.tick_to_price(tickLower / 2)
						currentTickRange = await this.pool.ticks(BigInt(tickLower))
					} else {
						// in the current price range
						let y = this.y_in_range(liquidity, sPriceCurrent, this.sPriceTarget)
						deltaTokens += y
						sPriceCurrent = this.sPriceTarget
					}
					console.log("token1 tradeSize: ", (deltaTokens / 10 ** this.token1.decimals), "from ", uniswapV3Exchange[this.exchange].protocol)
					// console.log("fixedNumber: ", deltaTokens.toFixed(this.token1.decimals))
					delta = BigInt(deltaTokens)
					return delta
				}
			}
		} catch (e: any) {
			if (e.code === 'BAD_DATA') {
				console.log("Protocol: ", uniswapV3Exchange[this.exchange].protocol)
				// console.log("Error: ", e)
				// return 0n
			}
			console.log('tickLower: ', tickLower)
			console.log(e.code)
		}
		return delta
	}
}

File: scripts/v3/modules/price/ref/calcVolToTarget.py
from web3 import Web3
from collections import namedtuple
# amount of x in range; sp - sqrt of current price, sb - sqrt of max price
def x_in_range(L, sp, sb):
    return L * (sb - sp) / (sp * sb)

# amount of y in range; sp - sqrt of current price, sa - sqrt of min price
def y_in_range(L, sp, sa):
    return L * (sp - sa)

def tick_to_price(tick):
    return 1.0001 ** tick

Tick = namedtuple("Tick", "liquidityGross liquidityNet feeGrowthOutside0X128 feeGrowthOutside1X128 tickCumulativeOutside secondsPerLiquidityOutsideX128 secondsOutside initialized")

# how much of X or Y tokens we need to *buy* to get to the target price?
deltaTokens = 0

if sPriceTarget > sPriceCurrent:
    # too few Y in the pool; we need to buy some X to increase amount of Y in pool
    while sPriceTarget > sPriceCurrent:
        if sPriceTarget > sPriceUpper:
            # not in the current price range; use all X in the range
            x = x_in_range(liquidity, sPriceCurrent, sPriceUpper)
            deltaTokens += x
            # query the blockchain for liquidity in the next tick range
            nextTickRange = Tick(*contract.functions.ticks(tickUpper).call())
            liquidity += nextTickRange.liquidityNet
            # adjust the price and the range limits
            sPriceCurrent = sPriceUpper
            tickLower = tickUpper
            tickUpper += tickSpacing
            sPriceLower = sPriceUpper
            sPriceUpper = tick_to_price(tickUpper // 2)
        else:
            # in the current price range
            x = x_in_range(liquidity, sPriceCurrent, sPriceTarget)
            deltaTokens += x
            sPriceCurrent = sPriceTarget
    print("need to buy {:.10f} X tokens".format(deltaTokens / 10 ** decimalsX))

elif sPriceTarget < sPriceCurrent:
    # too much Y in the pool; we need to buy some Y to decrease amount of Y in pool
    currentTickRange = None
    while sPriceTarget < sPriceCurrent:
        if sPriceTarget < sPriceLower:
            # not in the current price range; use all Y in the range
            y = y_in_range(liquidity, sPriceCurrent, sPriceLower)
            deltaTokens += y
            if currentTickRange is None:
                # query the blockchain for liquidityNet in the *current* tick range
                currentTickRange = Tick(*contract.functions.ticks(tickLower).call())
            liquidity -= currentTickRange.liquidityNet
            # adjust the price and the range limits
            sPriceCurrent = sPriceLower
            tickUpper = tickLower
            tickLower -= tickSpacing
            sPriceUpper = sPriceLower
            sPriceLower = tick_to_price(tickLower // 2)
            # query the blockchain for liquidityNet in new current tick range
            currentTickRange = Tick(*contract.functions.ticks(tickLower).call())
        else:
            # in the current price range
            y = y_in_range(liquidity, sPriceCurrent, sPriceTarget)
            deltaTokens += y
            sPriceCurrent = sPriceTarget
    print("need to buy {:.10f} Y tokens".format(deltaTokens / 10 ** decimalsY))


	
			let currentTickRange: ITick = {
				liquidityGross: 0n,
				liquidityNet: 0n,
				feeGrowthOutside0X128: 0n,
				feeGrowthOutside1X128: 0n,
				tickCumulativeOutside: 0n,
				secondsPerLiquidityOutsideX128: 0n,
				secondsOutside: 0n,
				initialized: false
			}

			let ATick = {
				liquidityTotal: 0n,
				liquidityDelta: 0n,
				outerFeeGrowth0Token: 0n,
				outerFeeGrowth1Token: 0n,
				prevTick: 0n,
				nextTick: 0n,
				outerSecondsPerLiquidity: 0n,
				outerSecondsSpent: 0n,
				hasLimitOrders: false,

			}